tensornetwork/backends/abstract_backend.py

```diff
@@ -429,6 +429,104 @@ def eigsh_lanczos(self,
     raise NotImplementedError(
         "Backend '{}' has not implemented eighs_lanczos.".format(self.name))
 
+  def gmres(self,
+            A_mv: Callable,
+            b: Tensor,
+            A_args: Optional[List] = None,
+            A_kwargs: Optional[dict] = None,
+            x0: Optional[Tensor] = None,
+            tol: float = 1E-05,
+            atol: Optional[float] = None,
+            num_krylov_vectors: Optional[int] = None,
+            maxiter: Optional[int] = 1,
+            M: Optional[Callable] = None
+            ) -> Tuple[Tensor, int]:
+    """ GMRES solves the linear system A @ x = b for x given a vector `b` and
+    a general (not necessarily symmetric/Hermitian) linear operator `A`.
+
+    As a Krylov method, GMRES does not require a concrete matrix representation
+    of the n by n `A`, but only a function
+    `vector1 = A_mv(vector0, *A_args, **A_kwargs)`
+    prescribing a one-to-one linear map from vector0 to vector1 (that is,
+    A must be square, and thus vector0 and vector1 the same size). If `A` is a
+    dense matrix, or if it is a symmetric/Hermitian operator, a different
+    linear solver will usually be preferable.
+
+    GMRES works by first constructing the Krylov basis
+    K = (x0, A_mv@x0, A_mv@A_mv@x0, ..., (A_mv^num_krylov_vectors)@x_0) and then
+    solving a certain dense linear system K @ q0 = q1 from whose solution x can
+    be approximated. For `num_krylov_vectors = n` the solution is provably exact
+    in infinite precision, but the expense is cubic in `num_krylov_vectors` so
+    one is typically interested in the `num_krylov_vectors << n` case.
+    The solution can in this case be repeatedly
+    improved, to a point, by restarting the Arnoldi iterations each time
+    `num_krylov_vectors` is reached. Unfortunately the optimal parameter choices
+    balancing expense and accuracy are difficult to predict in advance, so
+    applying this function requires a degree of experimentation.
+
+    In a tensor network code one is typically interested in A_mv implementing
+    some tensor contraction. This implementation thus allows `b` and `x0` to be
+    of whatever arbitrary, though identical, shape `b = A_mv(x0, ...)` expects.
+    Reshaping to and from a matrix problem is handled internally.
+
+    This function is supported only with the NumPy and Jax backends.
+    The numpy backend version is simply an interface to
```

Comment:
The last part can go, it's enough if it is described in the appropriate backend I think